File: SaidIt/src/test/java/eu/mrogalski/saidit/SaidItServiceTest.java
package eu.mrogalski.saidit;

import android.content.Context;
import android.content.SharedPreferences;
import androidx.test.core.app.ApplicationProvider;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.Shadows;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.android.controller.ServiceController;
import org.robolectric.annotation.LooperMode;
import org.robolectric.shadows.ShadowLooper;

import static org.junit.Assert.assertEquals;

@RunWith(RobolectricTestRunner.class)
@LooperMode(LooperMode.Mode.PAUSED)
public class SaidItServiceTest {

    private ServiceController<SaidItService> controller;
    private SaidItService saidItService;

    @Before
    public void setUp() {
        // Use Robolectric to create and manage the service lifecycle
        controller = Robolectric.buildService(SaidItService.class);
        saidItService = controller.create().get();
        saidItService.mIsTestEnvironment = true;
    }

    @After
    public void tearDown() {
        controller.destroy(); // Calls onDestroy and queues up thread termination
        if (saidItService.audioHandler != null && saidItService.audioHandler.getLooper() != null) {
            Shadows.shadowOf(saidItService.audioHandler.getLooper()).runToEndOfTasks(); // Process termination message
        }
        if (saidItService.audioThread != null) {
            try {
                // Wait for the thread to fully terminate
                saidItService.audioThread.join(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    private void runAllLoopers() {
        ShadowLooper.runUiThreadTasksIncludingDelayedTasks();
        if (saidItService.audioHandler != null && saidItService.audioHandler.getLooper() != null) {
            Shadows.shadowOf(saidItService.audioHandler.getLooper()).runToEndOfTasks();
        }
    }

    @Test
    public void testInitialState_startsListeningByDefault() {
        // By default, audio memory is enabled, so the service should start listening
        runAllLoopers();
        assertEquals(SaidItService.ServiceState.LISTENING, saidItService.state);
    }

    @Test
    public void testInitialState_isReadyWhenDisabled() {
        // Disable audio memory before the service is created
        Context context = ApplicationProvider.getApplicationContext();
        SharedPreferences prefs = context.getSharedPreferences(SaidIt.PACKAGE_NAME, Context.MODE_PRIVATE);
        prefs.edit().putBoolean(SaidIt.AUDIO_MEMORY_ENABLED_KEY, false).apply();

        // Re-create service to apply the preference
        controller.destroy(); // Destroy the old service instance
        controller = Robolectric.buildService(SaidItService.class);
        saidItService = controller.create().get();

        runAllLoopers();
        assertEquals(SaidItService.ServiceState.READY, saidItService.state);
    }

    @Test
    public void testEnableListening_changesState() {
        // Given the service is in the READY state
        saidItService.state = SaidItService.ServiceState.READY;

        // When listening is enabled
        saidItService.enableListening();
        runAllLoopers();

        // Then the state transitions to LISTENING
        assertEquals(SaidItService.ServiceState.LISTENING, saidItService.state);
    }

    @Test
    public void testDisableListening_changesState() {
        // Given the service is listening
        saidItService.enableListening();
        runAllLoopers();
        assertEquals(SaidItService.ServiceState.LISTENING, saidItService.state);

        // When listening is disabled
        saidItService.disableListening();
        runAllLoopers();

        // Then the state transitions back to READY
        assertEquals(SaidItService.ServiceState.READY, saidItService.state);
    }

    @Test
    public void testStartRecording_changesState() {
        // Given the service is listening
        saidItService.enableListening();
        runAllLoopers();
        assertEquals(SaidItService.ServiceState.LISTENING, saidItService.state);

        // When recording is started
        saidItService.startRecording(5.0f);
        runAllLoopers();

        // Then the state transitions to RECORDING
        assertEquals(SaidItService.ServiceState.RECORDING, saidItService.state);
    }

    @Test
    public void testStopRecording_changesState() {
        // Given the service is recording
        saidItService.enableListening();
        runAllLoopers();
        saidItService.startRecording(5.0f);
        runAllLoopers();
        assertEquals(SaidItService.ServiceState.RECORDING, saidItService.state);

        // When recording is stopped
        saidItService.stopRecording(null);
        runAllLoopers();

        // Then the state transitions back to LISTENING
        assertEquals(SaidItService.ServiceState.LISTENING, saidItService.state);
    }
}

File: SaidIt/src/test/kotlin/eu/mrogalski/saidit/SaidItServiceTest.kt
package eu.mrogalski.saidit

import android.app.AlarmManager
import android.content.Context
import android.net.Uri
import android.os.Build
import com.siya.epistemophile.audio.pcm.PcmAudioHelper
import com.siya.epistemophile.audio.pcm.WavAudioFormat
import com.siya.epistemophile.audio.pcm.WavFileWriter
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.Shadows.shadowOf
import org.robolectric.shadows.ShadowLooper
import java.io.File
import java.io.IOException
import kotlin.math.PI
import kotlin.math.sin

@OptIn(ExperimentalCoroutinesApi::class)
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [Build.VERSION_CODES.R])
class SaidItServiceTest {

    @get:Rule val mainDispatcherRule = MainDispatcherRule()

    @Test
    fun enableListening_reportsListeningStateViaCallback() = runTest(mainDispatcherRule.testDispatcher) {
        val controller = Robolectric.buildService(SaidItService::class.java)
        val service = controller.create().get()
        service.overrideAudioDispatcherForTest(mainDispatcherRule.testDispatcher)

        val memory = AudioMemory(FakeClock())
        service.overrideAudioMemoryForTest(memory)
        memory.allocate(AudioMemory.CHUNK_SIZE.toLong())
        fillAudioMemory(memory, generatePcmBytes(8_000, 1_600))

        service.isTestEnvironment = true
        service.enableListening()
        service.isTestEnvironment = false

        val captured = mutableListOf<ServiceState>()
        service.getState(object : SaidItService.StateCallback {
            override fun state(
                listeningEnabled: Boolean,
                recording: Boolean,
                bufferedSeconds: Float,
                totalSeconds: Float,
                recordedSeconds: Float
            ) {
                captured += ServiceState(listeningEnabled, recording, bufferedSeconds, totalSeconds, recordedSeconds)
            }
        })

        advanceUntilIdle()
        ShadowLooper.idleMainLooper()

        val state = captured.single()
        assertTrue(state.listeningEnabled)
        assertFalse(state.recording)
        assertTrue(state.bufferedSeconds > 0f)
        assertTrue(state.totalSeconds > 0f)
        controller.destroy()
    }

    @Test
    fun dumpRecording_persistsPcmBuffersToCache() = runTest(mainDispatcherRule.testDispatcher) {
        val controller = Robolectric.buildService(SaidItService::class.java)
        val service = controller.create().get()
        service.overrideAudioDispatcherForTest(mainDispatcherRule.testDispatcher)

        val memory = AudioMemory(FakeClock())
        service.overrideAudioMemoryForTest(memory)
        memory.allocate(AudioMemory.CHUNK_SIZE.toLong())
        val pcmBytes = generatePcmBytes(8_000, 4_000)
        fillAudioMemory(memory, pcmBytes)

        service.setSampleRate(8_000)
        service.audioSampleWriterFactory = { _, _, _, file -> FileBackedSampleWriter(file) }
        service.isTestEnvironment = true
        service.enableListening()
        service.isTestEnvironment = false

        val bytesPerSecond = (1f / service.getBytesToSeconds()).toInt()
        val memorySeconds = pcmBytes.size.toFloat() / bytesPerSecond + 0.05f
        val fileName = "unit_test_dump"

        service.dumpRecording(memorySeconds, null, fileName)

        advanceUntilIdle()
        ShadowLooper.idleMainLooper()

        val outputFile = File(service.cacheDir, "$fileName.m4a")
        assertTrue(outputFile.exists())
        val persisted = outputFile.readBytes()
        assertEquals(pcmBytes.size, persisted.size)
        assertTrue(pcmBytes.contentEquals(persisted))
        controller.destroy()
    }

    @Test
    fun dumpRecording_notifiesReceiverOnIoFailure() = runTest(mainDispatcherRule.testDispatcher) {
        val controller = Robolectric.buildService(SaidItService::class.java)
        val service = controller.create().get()
        service.overrideAudioDispatcherForTest(mainDispatcherRule.testDispatcher)

        val memory = AudioMemory(FakeClock())
        service.overrideAudioMemoryForTest(memory)
        memory.allocate(AudioMemory.CHUNK_SIZE.toLong())
        fillAudioMemory(memory, generatePcmBytes(8_000, 2_000))

        service.setSampleRate(8_000)
        val failure = IOException("expected failure")
        service.audioSampleWriterFactory = { _, _, _, _ -> FailingAudioSampleWriter(failure) }
        service.isTestEnvironment = true
        service.enableListening()
        service.isTestEnvironment = false

        var receivedError: Exception? = null
        val receiver = object : SaidItService.WavFileReceiver {
            override fun onSuccess(fileUri: Uri) = Unit
            override fun onFailure(e: Exception) { receivedError = e }
        }

        val bytesPerSecond = (1f / service.getBytesToSeconds()).toInt()
        val memorySeconds = 1.0f + (1f / bytesPerSecond)
        service.dumpRecording(memorySeconds, receiver, "failing_dump")

        advanceUntilIdle()
        ShadowLooper.idleMainLooper()

        assertNotNull(receivedError)
        assertEquals(failure, receivedError)
        controller.destroy()
    }

    @Test
    fun scheduleAutoSave_registersRepeatingAlarm() {
        val controller = Robolectric.buildService(SaidItService::class.java)
        val service = controller.create().get()

        val prefs = service.getSharedPreferences(SaidIt.PACKAGE_NAME, Context.MODE_PRIVATE)
        prefs.edit()
            .putBoolean("auto_save_enabled", true)
            .putInt("auto_save_duration", 120)
            .apply()

        service.scheduleAutoSave()

        val alarmManager = service.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val shadow = shadowOf(alarmManager)
        val scheduled = shadow.nextScheduledAlarm
        assertNotNull("Alarm should be scheduled", scheduled)
        scheduled!!
        assertEquals(AlarmManager.ELAPSED_REALTIME_WAKEUP, scheduled.type)
        assertEquals(120_000L, scheduled.interval)
        val pendingShadow = shadowOf(scheduled.operation)
        assertEquals(SaidItService::class.java.name, pendingShadow.savedIntent.component?.className)
        controller.destroy()
    }

    @Test
    fun enableListening_allocationFailureRevertsToReadyState() = runTest(mainDispatcherRule.testDispatcher) {
        val controller = Robolectric.buildService(SaidItService::class.java)
        val service = controller.create().get()
        service.overrideAudioDispatcherForTest(mainDispatcherRule.testDispatcher)

        val failingMemory = object : AudioMemory(FakeClock()) {
            var attempts = 0
            override fun allocate(sizeToEnsure: Long): Result<Unit> {
                attempts++
                return Result.failure(AudioMemoryException("failed to allocate"))
            }
        }
        service.overrideAudioMemoryForTest(failingMemory)

        service.isTestEnvironment = true
        service.disableListening()
        val initialState = service.stateForTest()
        service.isTestEnvironment = false
        service.enableListening()
        service.isTestEnvironment = true

        advanceUntilIdle()
        ShadowLooper.idleMainLooper()

        assertTrue("Expected at least one allocation attempt", failingMemory.attempts >= 1)
        assertEquals(initialState, service.stateForTest())
        controller.destroy()
    }

    private fun generatePcmBytes(sampleRate: Int, sampleCount: Int): ByteArray {
        val wavFormat = WavAudioFormat.mono16Bit(sampleRate)
        val samples = IntArray(sampleCount) { index ->
            (Short.MAX_VALUE * sin(2.0 * PI * index / sampleCount)).toInt()
        }
        val wavFile = newTempFile("input_${sampleRate}_$sampleCount", ".wav")
        WavFileWriter(wavFormat, wavFile).use { writer -> writer.write(samples) }
        val rawFile = newTempFile("input_${sampleRate}_$sampleCount", ".pcm")
        PcmAudioHelper.convertWavToRaw(wavFile, rawFile)
        return rawFile.readBytes()
    }

    private fun newTempFile(prefix: String, suffix: String): File {
        val sanitized = prefix.filter { it.isLetterOrDigit() || it == '_' || it == '-' }
        val finalPrefix = when {
            sanitized.length >= 3 -> sanitized
            sanitized.isNotEmpty() -> sanitized.padEnd(3, '0')
            else -> "saidit"
        }
        return File.createTempFile(finalPrefix, suffix).apply { deleteOnExit() }
    }

    private fun fillAudioMemory(memory: AudioMemory, pcmBytes: ByteArray) {
        var offset = 0
        memory.fill(AudioMemory.Consumer { array, arrayOffset, available ->
            if (offset >= pcmBytes.size) {
                return@Consumer Result.success(0)
            }
            val toCopy = minOf(available, pcmBytes.size - offset)
            System.arraycopy(pcmBytes, offset, array, arrayOffset, toCopy)
            offset += toCopy
            Result.success(toCopy)
        }).getOrThrow()
    }

    private data class ServiceState(
        val listeningEnabled: Boolean,
        val recording: Boolean,
        val bufferedSeconds: Float,
        val totalSeconds: Float,
        val recordedSeconds: Float
    )
}
